DefensiveCoding.md

```diff
@@ -0,0 +1,35 @@
+# Explanations for defensive coding
+
+## Immutability
+
+The concept that fields and properties cannot change once the object or struct has been created.
+This is achieved by marking all fields readonly, all properties get only and using the IReadOnly interfaces for collections and arrays.
+The Record type (from C# 9.0 and onwards) is immutable by default.
+The price for immutability is extensive copying and added code in the shape of With methods.
+
+See: https://en.wikipedia.org/wiki/Immutable_object
+
+## Parse, don't validate
+
+The concept that we parse input as soon as we receive it instead of validating same input all over the place.
```

Comment:
Perhaps the concept should specify that we're parsing into the type with the strongest constraints to ensure that we don't need repeated checks (or weakened output constraints later on). Maybe also give a stricter example than just programming by name - but something like asserting that a product id is always 8 characters long containing only alphanumeric characters.